### PR TITLE
[WIP] added password argument to encrypter

### PR DESCRIPTION
**Pull request by Jameron**
I am building an application that requires user input password encrypted data. Currently Laravel defaults to using the APP_KEY as the encryption password. I added updates that will allow users to pass in a custom password string to the encrypt, encryptString, decrypt, and decryptString methods. If no password, the default APP_KEY will be used. This update will allow Laravel users to easily add custom passwords to their encrypted data. 

In an age of domestic surveillance and data warehousing, it is crucial that developers can easily access the encryption tools needed to provide users secure data.

### CODE REVIEW

**Comment by sisve** [comment]
1. You've forgot to update the contract/interface.
2. I would prefer if the parameter would be renamed to "key" instead of password, because that is what you're sending in.
3. I would prefer to see a separate method for this. I do not want any "default" behavior if the passed $key was null; I would expect an exception.
4. Shouldn't you be using some key-expansion techniques to build the key from the user-provided data?

**Comment by Jameron** [comment]
1) Thanks, good catch. I have updated the contract.
2) The encrypt and decrypt methods are using openssl_encrypt/decrypt under the hood. The 3rd parameter is titled password in the documentation. Please confirm: http://php.net/manual/en/function.openssl-encrypt.php
3) Currently by default Laravel sets the encryption password as the APP_KEY.
4) I'm not sure, from my understanding it is only a password. If you have some supporting documentation or link to help me out please share.

File: src/Illuminate/Contracts/Encryption/Encrypter.php [hunk 1]
```diff
@@ -11,7 +11,7 @@
      * @param  bool  $serialize
      * @return string
      */
-    public function encrypt($value, $serialize = true);
+    public function encrypt($value, $serialize = true, $password = null);
```
Comment (lucasmichot) [inline src/Illuminate/Contracts/Encryption/Encrypter.php]: I would use `$key` too as per @sisve comment.
Docblock?

File: src/Illuminate/Contracts/Encryption/Encrypter.php [hunk 2]
```diff
@@ -20,5 +20,5 @@ public function encrypt($value, $serialize = true);
      * @param  bool  $unserialize
      * @return string
      */
-    public function decrypt($payload, $unserialize = true);
+    public function decrypt($payload, $unserialize = true, $password = null);
```
Comment (lucasmichot) [inline src/Illuminate/Contracts/Encryption/Encrypter.php]: Docblock?

File: src/Illuminate/Encryption/Encrypter.php [hunk 3]
```diff
@@ -68,7 +68,7 @@ public static function supported($key, $cipher)
      *
      * @throws \Illuminate\Contracts\Encryption\EncryptException
      */
-    public function encrypt($value, $serialize = true)
+    public function encrypt($value, $serialize = true, $password = null)
```
Comment (lucasmichot) [inline src/Illuminate/Encryption/Encrypter.php]: Docblock?

File: src/Illuminate/Encryption/Encrypter.php [hunk 4]
```diff
@@ -77,7 +77,7 @@ public function encrypt($value, $serialize = true)
         // value can be verified later as not having been changed by the users.
         $value = \openssl_encrypt(
             $serialize ? serialize($value) : $value,
-            $this->cipher, $this->key, 0, $iv
+            $this->cipher, ($password) ? $password : $this->key, 0, $iv
```
Comment (lucasmichot) [inline src/Illuminate/Encryption/Encrypter.php]: `! is_null($password)`

File: src/Illuminate/Encryption/Encrypter.php [hunk 5]
```diff
@@ -104,9 +104,9 @@ public function encrypt($value, $serialize = true)
      * @param  string  $value
      * @return string
      */
-    public function encryptString($value)
+    public function encryptString($value, $password = null)
```
Comment (lucasmichot) [inline src/Illuminate/Encryption/Encrypter.php]: Docblock

File: src/Illuminate/Encryption/Encrypter.php [hunk 6]
```diff
@@ -118,7 +118,7 @@ public function encryptString($value)
      *
      * @throws \Illuminate\Contracts\Encryption\DecryptException
      */
-    public function decrypt($payload, $unserialize = true)
+    public function decrypt($payload, $unserialize = true, $password = null)
```
Comment (lucasmichot) [inline src/Illuminate/Encryption/Encrypter.php]: Docblock

File: src/Illuminate/Encryption/Encrypter.php [hunk 7]
```diff
@@ -128,7 +128,7 @@ public function decrypt($payload, $unserialize = true)
         // we will then unserialize it and return it out to the caller. If we are
         // unable to decrypt this value we will throw out an exception message.
         $decrypted = \openssl_decrypt(
-            $payload['value'], $this->cipher, $this->key, 0, $iv
+            $payload['value'], $this->cipher, ($password) ? $password : $this->key, 0, $iv
```
Comment (lucasmichot) [inline src/Illuminate/Encryption/Encrypter.php]: `! is_null($password)`

File: src/Illuminate/Encryption/Encrypter.php [hunk 8]
```diff
@@ -144,9 +144,9 @@ public function decrypt($payload, $unserialize = true)
      * @param  string  $payload
      * @return string
      */
-    public function decryptString($payload)
+    public function decryptString($payload, $password = null)
```
Comment (lucasmichot) [inline src/Illuminate/Encryption/Encrypter.php]: Docblock?

File: tests/Encryption/EncrypterTest.php [hunk 9]
```diff
@@ -15,6 +15,14 @@ public function testEncryption()
         $this->assertEquals('foo', $e->decrypt($encrypted));
     }
 
+    public function testEncryptionWithPassword()
+    {
+        $e = new Encrypter(str_repeat('a', 16));
+        $encrypted = $e->encrypt('foo','bar');
+        $this->assertNotEquals('foo', $encrypted);
+        $this->assertEquals('foo', $e->decrypt($encrypted,'bar'));
```
Comment (lucasmichot) [inline tests/Encryption/EncrypterTest.php]: `assertSame`

File: tests/Encryption/EncrypterTest.php [hunk 10]
```diff
@@ -23,6 +31,14 @@ public function testRawStringEncryption()
         $this->assertEquals('foo', $e->decryptString($encrypted));
     }
 
+    public function testRawStringEncryptionWithPassword()
+    {
+        $e = new Encrypter(str_repeat('a', 16));
+        $encrypted = $e->encryptString('foo','bar');
+        $this->assertNotEquals('foo', $encrypted);
+        $this->assertEquals('foo', $e->decryptString($encrypted,'bar'));
```
Comment (lucasmichot) [inline tests/Encryption/EncrypterTest.php]: `assertSame`

**Comment by sisve** [comment]
Regarding openssl_encrypt, the description is "Encrypts given data with given method and _key_". The comments include "Hash the chosen _encryption key_ (the password parameter)", "password - As $username mentioned before, this is actually _THE KEY!_", "password - BINARY string (ie. the _encryption key_ in binary)". My point is that it's an encryption key, and that openssl_encrypt has a badly named parameter. We should not that leak over into your abstraction, the Encrypter.

Your third point doesn't really answer anything. I'm arguing that someone doing `$encrypter->encrypt($value, true, $customKey)` should get an exception if the $customKey is empty; it should not silently default to another key ($this->key).

You cannot just send in any random string as a new key, it has to match the cipher used. See https://github.com/laravel/framework/blob/5.4/src/Illuminate/Encryption/Encrypter.php#L54-L60 for examples. This means that you have to have some way to transform user input ("hello world") into a key of appropriate length (16 or 32 bytes) and characteristics (randomness-ish). This is what key expansion/stretching is.

Relevant links:
* https://secure.php.net/manual/en/function.hash-pbkdf2.php
* https://en.wikipedia.org/wiki/Key_derivation_function

**Comment by Jameron** [comment]
Yes, after posting my response I did some more reading. I realize now that password is a misnomer and key is more appropriate.

I also now see your reasoning for separating this out into a separate method, a null $key argument should throw an exception.

I will need some time to work on this tomorrow, hopefully it is possible and I can find a way to convert a string 'swordfish' to an acceptable key format. I am now concerned that what I am trying to do may not be possible.

**Comment by ntzm** [comment]
Changing the contracts is a breaking change. If you want to change the contract it should target 5.5

**Comment by Jameron** [comment]
I didn't see a 5.5 branch, how do I target 5.5? Also, I will be working on this some more this evening, please keep open until I know if this is possible.

**Comment by fernandobandeira** [comment]
5.5 is the master branch 😄 if you'll be working on it change the title including the [WIP] in the beginning...

**Comment by Jameron** [comment]
Shoot looks I need to move this to a new pull request to switch to master?

**Pull request closed without merging**